// @neondatabase/auth/dist/constants-2bpp2_-f.mjs
//#region src/core/constants.ts
/**
* Session caching configuration constants
*
* Uses industry-standard 60s cache TTL (common across auth providers).
*
* Note: Token refresh detection is now automatic via Better Auth's
* fetchOptions.onSuccess callback. No polling is needed.
*/
/** Session cache TTL in milliseconds (60 seconds) */
const SESSION_CACHE_TTL_MS = 6e4;
/** Clock skew buffer for token expiration checks in milliseconds (10 seconds) */
const CLOCK_SKEW_BUFFER_MS = 1e4;
/** Default session expiry duration in milliseconds (1 hour) */
const DEFAULT_SESSION_EXPIRY_MS = 36e5;
/** Name of the session verifier parameter in the URL, used for the OAUTH flow */
const NEON_AUTH_SESSION_VERIFIER_PARAM_NAME = "neon_auth_session_verifier";
/** Name of the popup marker parameter in the URL, used for OAuth popup flow in iframes */
const NEON_AUTH_POPUP_PARAM_NAME = "neon_popup";
/** Name of the original callback URL parameter, used in OAuth popup flow */
const NEON_AUTH_POPUP_CALLBACK_PARAM_NAME = "neon_popup_callback";
/** The callback route used for OAuth popup completion (must be in middleware SKIP_ROUTES) */
const NEON_AUTH_POPUP_CALLBACK_ROUTE = "/auth/callback";
/** Message type for OAuth popup completion postMessage */
const OAUTH_POPUP_MESSAGE_TYPE = "neon-auth:oauth-complete";
/** Prefix for all Neon Auth cookies */
const NEON_AUTH_COOKIE_PREFIX = "__Secure-neon-auth";

//#endregion
export { NEON_AUTH_POPUP_CALLBACK_ROUTE as a, OAUTH_POPUP_MESSAGE_TYPE as c, NEON_AUTH_POPUP_CALLBACK_PARAM_NAME as i, SESSION_CACHE_TTL_MS as l, DEFAULT_SESSION_EXPIRY_MS as n, NEON_AUTH_POPUP_PARAM_NAME as o, NEON_AUTH_COOKIE_PREFIX as r, NEON_AUTH_SESSION_VERIFIER_PARAM_NAME as s, CLOCK_SKEW_BUFFER_MS as t };
